### ISSUE TITLE
Training encoder from deliminator loss?

### ISSUE THREAD

**Issue by vintersnow** [zhengliz/natural-adversary#2]
Hi, thank you for sharing the code. I read your paper and I am trying to train and modify.
I have a question about how you optimize the encoder.
You are optimizing the encoder and the critic from the same loss at `train_gan_d`, but is this collect?
In my understanding, the encoder's loss function and the critic loss function are different.

![2018-07-13-125133_715x69_scrot](https://user-images.githubusercontent.com/6247314/42671563-901786ca-869b-11e8-942a-efdd8bd2d3e7.png)

**Comment by zhengliz** [comment]
Actually, the critic learns to maximize the function, while others to minimize.

**Comment by vintersnow** [comment]
OK, I understand. You are flipping the gradients using `grad_hook`.
Thank you, and sorry for bothering.
